src/slice/new_iter.rs

```diff
@@ -0,0 +1,495 @@
+		unsafe {
+			// Addition safety guarantees are described three lines higher.
+			let plus_rear: usize = unchecked_add!(
+				bits_len,
+				*O::idx_from_sel($self.bit_end) as usize
+			);
+			// SAFETY: `bit_end` is always >= `bit_ptr` if pointers are equal
+			// by the type invariant.
+			// If pointers aren't eq, `bits_len` is always bigger than `cttz(bit_ptr)`;
+			// and so `plus_rear` is always >= `bits_len`.
+			unchecked_sub!(plus_rear, *O::idx_from_sel($self.bit_ptr) as usize)
+			}
+		}};
+}
+
+// Shifts `bit_ptr` to the left by one and returns the bit at old position as bool.
+// It is not unsafe by itself, but when called onto `bit_ptr` with only last bit set,
+// it *could* wreck havoc.
+macro_rules! next_bit_unchecked {
+	($self:ident, $store:expr) => {{
+		let old = $self.bit_ptr & $store != 0;
+		$self.bit_ptr <<= 1;
+		old
+		}};
+}
+
+// Rotates `bit_end` to the right by one and returns the bit at new position as bool.
+// It is the best solution I could come up with.
+macro_rules! next_back_bit_unchecked {
+	($self:ident, $store:expr, $t:ty) => {{
+		// $self.bit_end = $self.bit_end.rotate_right(1);
+		$self.bit_end = O::rot_backward($self.bit_end, 1);
+		*$self.bit_end & $store != <$t>::ZERO
+		}};
+}
+
+// // Offsets `ptr` pointer by 1 and sets `bit_ptr` to the FIRST_BIT.
+// No more. Now this rotates too and need no setting.
+macro_rules! forward_ptr_unchecked {
+	($self:ident) => {{
+		// $self.bit_ptr = <$t as BitMemory>::ONE;
+		$self.ptr = NonNull::new_unchecked($self.ptr.as_ptr().offset(1));
+		}};
+}
+
+// Offsets `end` pointer by -1. Yes, that's all.
+// That's 'cuz `bit_end` rotates, not shifts.
+macro_rules! backward_end_unchecked {
+	($self:ident) => {{
+		$self.end = $self.end.offset(-1);
+		}};
+}
+
+impl<'a, T: BitMemory, O: BitOrder> Iter<'a, T, O> {
+	// /// Number of bits in the `$store_type`. Very useful for many bit operations.
+	// const BITS: u8 = mem::size_of::<T>() as u8 * 8;
+
+	/// Constant with only last bit set.
+	///
+	/// It could've been simple `1.rotate_right(1)`,
+	/// but stabilized version can't infer type.
+	///
+	/// This could never overflow when used with numeric primitives,
+	/// because `T` could always hold `size_of::<Self>() * 8`.
+	/// This could underflow with ZST, but ZSTs don't work as bitstores.
+	// const LAST_BIT: T = T::ONE << (Self::BITS - 1);
+	// const LAST_BIT: T = T::ONE << T::MASK;
+	// Ffuuuuuu... I can't do this without intri.
+	// const LAST_BIT: T = intrinsics::bitreverse(T::ONE);
+
+	// /// Returns true if the iterator is empty. This is faster than `[`len()`] == 0`
+	// ///
+	// /// [`len()`]: https://doc.rust-lang.org/std/iter/trait.ExactSizeIterator.html#method.len
+	// #[inline]
+	// pub fn is_empty(&self) -> bool {
+	// 	is_empty!(self)
+	// }
+
+	#[inline]
+	fn start_bit() -> BitSel<T> {
+		O::at(unsafe { BitIdx::new_unchecked(0) }).select()
+	}
+
+	#[inline]
+	fn last_bit() -> BitSel<T> {
+		O::at(unsafe { BitIdx::new_unchecked(T::BITS - 1) }).select()
```

Comment:
I'll need to document that `BitMemory::MASK` is the constant you're computing here

Reply:
I know, but I have problems with it's name, as it's not really clear that `MASK` is `BITS - 1`.

Reply:
That's good feedback! I will freely admit that I chose `MASK` over `MAX` because when I named it I was using it for `& val` clears more often than `< val` tests. The doc comment above it is a pretty clear indicator that it should be `MAX`, though.

Reply:
> that it should be `MAX`

More like a `MAX_INDEX` than simple `MAX`.